### ISSUE TITLE
Request: Expose callback for when Drawer finishes closing

### ISSUE THREAD

**Issue by brianwilliams42** [flutter/flutter#14510]
## Steps to Reproduce

I'm attempting to take a screenshot of the app to attach to a user-submitted feedback report. The "Send feedback" button for our app lives in the Material Drawer that slides out from the side of the main app Scaffold. But users typically aren't sending feedback about that Drawer, they are sending feedback about the view underneath the Drawer.

Here's my ideal behavior:
- Open Drawer
- Tap "Send Feedback" item
- Drawer-closing animation begins (still covering most of the underlying content)
- some time passes...
- Drawer animation ends (Drawer now gone, previous view exposed)
- Trigger screenshot and create feedback request

If you have an Android phone with the Gmail app, you can see how I envision this working. I couldn't find an iOS example that wasn't a screenshot of the drawer (Google Calendar screenshots the drawer).

I can't figure out the best way to know when the Drawer is hidden, currently my screenshots trigger immediately and just capture the Drawer. 

A callback may not be the best solution, but was my naive first instinct. If there was a fancy way to "screenshot" the underlying widgets without waiting for the Drawer to hide, that'd be slick, but probably far more complex than necessary.

**Comment by zoechi** [comment]
#30856 seems to address that

**Comment by Hixie** [comment]
The feature described here is about taking a screenshot of the app at some point in the future (with the drawer closed).

@brianwilliams42 can you talk more about your needs? e.g. how do you want to handle the case of the user tapping "send feedback" then holding the drawer open as it closes? or the user quickly tapping some other button that triggers a dialog or some such?

@Jon-Salmon @msyaojz can you talk more about your use case for this feature?

**Comment by Hixie** [comment]
I added some comments in #20856 about possible ways to address this. I'm leaning towards exposing the animation controllers (as `Animation<double>s`), but I don't know if that addresses all the use cases, hence the questions above.

**Comment by Jon-Salmon** [comment]
Thanks for taking a look at this, my use case is as follows:

I'm producing an app for displaying guitar music which requires both the ability to measure the exact width of rendered text prior to building the view and also must be able to have very smooth scrolling of deep widget trees.
As flutter doesn't support the former (please correct me if i'm wrong) and I couldn't accheive the performance I wanted of the latter, my solution has been to use a flutter view as an overlay over a native android app. A transparent section in the flutterView reveials the android app below that displays the acutal music.

This works really well (although I'm aware its not techncally a supported feature of flutter currently) with the only slight issue relating to touch events.

In a nutshell, I need to be able to tell when a drawer is opens/closes so as to disable/reenable touch events on the android part of the app to stop the andoird part scrolling while scrolling a listview in the drawer.

Ideally, being able to detect when the drawer starts to open or close would be best.

A beta version of the app can be found [here](https://play.google.com/store/apps/details?id=com.songbookpro.songbookpro) incase this makes no sense.

**Comment by rostopira** [comment]
This issue is almost one year old, any progress?

**Comment by bosphere** [comment]
I'd like to share another use case for this request and hopefully it'll justify the expected level of urgency here.

It's quite a common design these days to have the status bar completely transparent to have that immersive feeling of the content behind. And oftentimes for better readability the status bar text needs to be rendered either in a light colour or in a dark colour. But that colour isn't always a perfect fit to the drawer's background of chosen.

We need this callback so that many things can be done when the drawer is open or not i.e. changing the status bar text colour.

**Comment by Hixie** [comment]
@Jon-Salmon We support measuring text, see the `TextPainter` API.

@bosphere For changing the colour of the notification area, I recommend using the technique described at the bottom of https://master-docs.flutter.io/flutter/services/SystemChrome/setSystemUIOverlayStyle.html (using the `AnnotatedRegion<SystemUiOverlayStyle>`).

**Comment by bosphere** [comment]
@Hixie Thanks a ton for pointing me to the right direction! `AnnotatedRegion` works perfectly for me to toggle between system UI styles.

**Comment by DelmarZhao** [comment]
Are there any updates on this feature? I am in a scenario where My drawer is being covered by a flushbar. The creator of flushbar said that since Flushbar pushes a new route, it will be drawn over my drawer, so I am looking to detect when the drawer is being opened so that I may hide the flushbar. 

Is there perhaps another way to make my drawer draw on top of the flushbar?

**Comment by intako** [comment]
I also have a case, when I'd like to have this feature. I need to make requests when drawer's been open. But we can open a drawer by swiping, so making custom burger menu icon with handler doesn't work for me.

**Comment by chunhtai** [comment]
> Are there any updates on this feature? I am in a scenario where My drawer is being covered by a flushbar. The creator of flushbar said that since Flushbar pushes a new route, it will be drawn over my drawer, so I am looking to detect when the drawer is being opened so that I may hide the flushbar.
> 
> Is there perhaps another way to make my drawer draw on top of the flushbar?

I am not quite understand the use case, If flush bar covers the drawer, how would you close it in the first place?

**Comment by Hixie** [comment]
@intako can you talk more about this use case? Why would you want to trigger a request?

**Comment by grandpa-guru** [comment]
@Hixie Another use case: I have an analytics app and all the filters are in the drawer. User can modify all filters as he likes and when drawer closes we want to refresh the charts. That callback is essential for mentioned use case.

**Comment by grandpa-guru** [comment]
> @Hixie Another use case: I have an analytics app and all the filters are in the drawer. User can modify all filters as he likes and when drawer closes we want to refresh the charts. That callback is essential for mentioned use case.

Possible workaround that works for me: refreshing charts in drawer dispose method.

**Comment by valterh4ck3r** [comment]
Waiting for workaround.

**Comment by chunhtai** [comment]
@valterh4ck3r can you provide your use case? that will help us come up with a better approach that can solve all use case

**Comment by anisalibegic** [comment]
@chunhtai https://github.com/flutter/flutter/issues/14510#issuecomment-532089130

**Comment by valterh4ck3r** [comment]
I need that on close drawer, update ui.

**Comment by chunhtai** [comment]
will get back to this once i have time

**Comment by valterh4ck3r** [comment]
thanks

**Comment by AlexHartford** [comment]
This should be resolved by https://github.com/flutter/flutter/pull/67249, when can this be added to stable?

**Comment by bikcrum** [comment]
> This should be resolved by #67249, when can this be added to stable?

It has been already released in stable version 2.0.0

Thank you.
